content/posts/2018-01-10-kotlin-null.md

```diff
@@ -0,0 +1,249 @@
+if a return value or an argument can be `null`...
+The situation becomes even worse with non-library and non-open source projects with deadlines and priorities.
+
+### Habits
+
+I’ve spent something about 1.5 years writing Kotlin-only code every single workday.
+Sometimes Python, sometimes Bash, but it does not change the picture.
+
+The human mindset adapts very well. Typing Kotlin code and thinking with Kotlin concepts every day
+_will_ make you forget about simple things — like the one we are discussing here — Java does not have
+nullability metadata out of the box. You will expect that everything you touch actually has it
+(a classic [Midas mistake](https://en.wikipedia.org/wiki/Midas#Golden_Touch)),
+because your shiny Kotlin codebase has it and you have a great confidence in that.
+
+Prepare for unforeseen consequences.
+
+## The Solution
+
+### The Obvious
+
+This is a hard one since it is based on a non-automated semantics mechanism — the mind.
+It is always better to remember when you are using Java code and when you are using the Kotlin one.
+
+Ask more _strange_ questions during a code review — _Are we guarded against this Java code?_
+It may sound dumb and people are often too lazy to check but it is better to raise a question instead of doing nothing.
+
+### Contributions
+
+Let’s face it — a lot of code we use as libraries is open source.
+The entire Android Framework is here — clone it, read it, modify it.
+Yes, it is not so easy to talk people into using nullability annotations,
+especially if your main argument is “I use Kotlin, so you should...”.
+
+* Maintainers are not your personal minions, so please show at least some respect.
+* The Kotlin argument does not work all the time and it should be the last one you have up your sleeve.
+  Not so many people actually use Kotlin at this point, but there is a huge amount of Java code.
+* Nullability annotations improve documentation and IDE support for everyone who uses the Java code,
+  so it seems like a better argument.
+* Pull Requests are always welcome!
+
+### Facade the Hell of It
+
+The worst case scenario — you do not have access to the source code, maintainers are quiet for years.
+
+Make your own abstraction using Kotlin! Believe me, almost everything is doable with enough effort
+(excluding codebases with a huge number of public methods — like RxJava).
+As a side benefit — you will make your code more testable encapsulating implementation behind an interface.
+
+Sample time!
+
+```java
+class CorporateCalculator {
+    void calculate(Argument argument) {
+        if (argument == null) throw new NullPointerException("Gotcha!");
+
+        calculateUsingCorporateBlackMagic(argument);
+    }
+}
+```
+```kotlin
+interface Calculator {
+    fun calculate(argument: Argument)
+
+    class CorporateImpl : Calculator {
+        override fun calculate(argument: Argument) {
+            CorporateCalculator().calculate(argument)
```

Comment:
Just a note: this will duplicate == null check

Reply:
Not sure I understand you correctly.

The Kotlin interface will disallow passing `null`, and yes, the Java class will check for `null` under the hood as well.